src/main/java/org/killbill/billing/plugin/adyen/client/notification/AdyenNotificationService.java

```diff
@@ -64,63 +64,59 @@ public String handleNotifications(final String input) {
-
-        SendNotification sendNotification = null;
+        final SendNotification sendNotification;
         try {
             sendNotification = parse(inputStream);
         } catch (final Exception e) {
             logger.warn("Error parsing Adyen notification", e);
+            return createSendNotificationResponse("error");
         }
 
-        if (sendNotification != null) {
-            final List<NotificationRequestItem> listOfNotifications = sendNotification.getNotification()
-                                                                                      .getNotificationItems()
-                                                                                      .getNotificationRequestItem();
-
-            for (final NotificationRequestItem item : listOfNotifications) {
-                try {
-                    handleNotification(item);
-                } catch (final Exception e) {
-                    logger.warn("Error handling notification: eventCode='{}', pspReference='{}', originalReference='{}', success='{}', reason='{}', merchantReference='{}', merchantAccount='{}'",
-                                item.getEventCode(),
-                                item.getPspReference(),
-                                item.getOriginalReference(),
-                                item.isSuccess(),
-                                item.getReason(),
-                                item.getMerchantReference(),
-                                item.getMerchantAccountCode(),
-                                e);
-                }
-            }
-            response = "[accepted]";
-        }
-
-        try {
-            return createSendNotificationResponse(response);
-        } catch (final Exception e) {
-            // Not much we can do. Let Adyen retry.
-            throw new RuntimeException(e);
+        final List<NotificationRequestItem> listOfNotifications = sendNotification.getNotification()
+                                                                                  .getNotificationItems()
+                                                                                  .getNotificationRequestItem();
+        for (final NotificationRequestItem item : listOfNotifications) {
+            handleNotification(item);
         }
+        return createSendNotificationResponse("[accepted]");
     }
 
     private void handleNotification(final NotificationRequestItem item) {
-        logger.info("Handling notification: eventCode='{}', pspReference='{}', originalReference='{}', success='{}', reason='{}', merchantReference='{}', merchantAccount='{}'",
+        Exception error = null;
+        final long startTime = System.currentTimeMillis();
+        long duration = 0L;
+        try {
+            final AdyenNotificationHandler adyenNotificationHandler = getAdyenNotificationHandler(item);
+            if (adyenNotificationHandler == null) {
+                logger.warn("No handler available - ignoring");
```

Comment:
Not sure if it's worth it. Current implementation of `canHandleNotification` always returns true, so this should never happen.

To be honest, I'm not even sure we need whole "factoryish" method thing, but that's a different refactor.